lib/spells/effects/Obstacle.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Obstacle.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include "LocationEffect.h"
+#include "../../battle/BattleHex.h"
+
+namespace spells
+{
+namespace effects
+{
+
+class ObstacleSideOptions
+{
+public:
+	using RelativeShape = std::vector<std::vector<BattleHex::EDir>>;
+
+	RelativeShape shape; //shape of single obstacle relative to obstacle position
+	RelativeShape range; //position of obstacles relative to effect destination
+
+	std::string appearAnimation;
+	std::string animation;
+
+	int offsetY;
+
+	ObstacleSideOptions();
+
+	void serializeJson(JsonSerializeFormat & handler);
+
+private:
+	void serializeRelativeShape(JsonSerializeFormat & handler, const std::string & fieldName, RelativeShape & value);
+};
+
+class Obstacle : public LocationEffect
+{
+public:
+	Obstacle(const int level);
+	virtual ~Obstacle();
+
+	void adjustAffectedHexes(std::set<BattleHex> & hexes, const Mechanics * m, const Target & spellTarget) const override;
+
+	bool applicable(Problem & problem, const Mechanics * m) const override;
+	bool applicable(Problem & problem, const Mechanics * m, const Target & aimPoint, const EffectTarget & target) const override;
+
+	EffectTarget transformTarget(const Mechanics * m, const Target & aimPoint, const Target & spellTarget) const override;
+
+protected:
+	void apply(BattleStateProxy * battleState, RNG & rng, const Mechanics * m, const EffectTarget & target) const override;
+	void serializeJsonEffect(JsonSerializeFormat & handler) override;
+
+private:
+	bool hidden;
+	bool passable;
+	bool trigger;
+	bool trap;
```

Comment:
These bools looks ideal like in SpellCreatedObstacle. Maybe there is some way to reduce that code?

Reply:
I could unify ObstacleSideOptions with CObstacleInstance by f.e. common base class, but code will be less readable.